### PR TITLE
Prove possession of a credential

### PR DESCRIPTION
**Pull request by miha-stopar**
User can now prove the possession of a credential to organization.

### CODE REVIEW

File: crypto/zkp/schemes/cl/org_credential_issuer.go [hunk 1]
```diff
@@ -251,4 +242,4 @@ func (i *OrgCredentialIssuer) UpdateCredential(nonceUser *big.Int, newKnownAttrs
 	i.Org.receiverRecords[i.nym] = receiverRecord
 
 	return NewCredential(newA, e, v11), AProof
-}
+}
```
Comment (mancabizjak) [inline crypto/zkp/schemes/cl/org_credential_issuer.go]: Please add newline.

File: crypto/zkp/schemes/cl/user.go [hunk 2]
```diff
@@ -25,22 +25,31 @@ import (
 
 	"github.com/xlab-si/emmy/crypto/commitments"
 	"github.com/xlab-si/emmy/crypto/common"
+	"github.com/xlab-si/emmy/crypto/groups"
+	"github.com/xlab-si/emmy/crypto/zkp/primitives/qrspecialrsaproofs"
 )
 
+// User should be created when a credential is to be issued, updated or proved. If a new credential under
+// a new nym is needed, new User instance is needed.
 type User struct {
```
Comment (mancabizjak) [inline crypto/zkp/schemes/cl/user.go]: I think we're missing `goimports` formatting.

File: crypto/zkp/schemes/cl/user_credential_receiver.go [hunk 3]
```diff
@@ -254,15 +249,15 @@ func (r *UserCredentialReceiver) VerifyCredential(cred *Credential,
...
 	if AProof.Challenge.Cmp(c) != 0 {
 		return false, fmt.Errorf("challenge is not correct")
 	}
 
-	credentialVerifier.SetChallenge(AProof.Challenge)
-	return credentialVerifier.Verify(AProof.ProofData), nil
-}
+	ver.SetChallenge(AProof.Challenge)
+	return ver.Verify(AProof.ProofData), nil
+}
```
Comment (mancabizjak) [inline crypto/zkp/schemes/cl/user_credential_receiver.go]: Please add newline + `goimports` formatting.

File: crypto/zkp/schemes/cl/user.go [hunk 4]
```diff
@@ -85,33 +93,123 @@ func NewUser(clParams *CLParams, clPubKey *PubKey, pedersenParams *commitments.P
...
+	denomInv := group.Inv(denom)
+	y := group.Mul(u.PubKey.Z, denomInv)
+
+	prover := qrspecialrsaproofs.NewRepresentationProver(group, u.Params.SecParam,
+		secrets, bases, y)
+
+	// boundary for m_tilde
+	b_m := u.Params.AttrBitLen + u.Params.SecParam + u.Params.HashBitLen
+	// boundary for e1
+	b_e := u.Params.EBitLen + u.Params.SecParam + u.Params.HashBitLen
```
Comment (mancabizjak) [inline crypto/zkp/schemes/cl/user.go]: Would you mind renaming `b_e` to `b_e1` for consistency?

**Comment by miha-stopar** [comment]
Thanks Manca! Goimport and gofmt applied. For GetContext - it was not intentional, I fixed it now (it didn't affect the execution of a program or security though). For b_e - the mistake was in the comment.

**Pull request merged**